staging/src/k8s.io/client-go/plugin/pkg/client/auth/exec/exec.go

```diff
@@ -353,40 +363,62 @@ func (r *roundTripper) RoundTrip(req *http.Request) (*http.Response, error) {
 			Code:   int32(res.StatusCode),
 		}
 		if err := r.a.maybeRefreshCreds(creds, resp); err != nil {
-			klog.Errorf("refreshing credentials: %v", err)
+			klog.ErrorS(err, "refreshing credentials")
 		}
 	}
 	return res, nil
 }
 
-func (a *Authenticator) credsExpired() bool {
+// contextKey type is unexported to prevent collisions.
+type contextKey int
+
+const (
+	roundTripKey contextKey = iota
+	certKey
+)
+
+func (a *Authenticator) credsExpiredLocked() bool {
 	if a.exp.IsZero() {
 		return false
 	}
 	return a.now().After(a.exp)
 }
 
-func (a *Authenticator) cert() (*tls.Certificate, error) {
-	creds, err := a.getCreds()
+func (a *Authenticator) cert(ctx context.Context) (*tls.Certificate, error) {
+	// ensure a single exec plugin call per RoundTrip
+	if didRoundTrip, _ := ctx.Value(roundTripKey).(bool); didRoundTrip {
+		cert, _ := ctx.Value(certKey).(*tls.Certificate)
+		return cert, nil
+	}
+
+	// fallback to fetching creds directly if we are not called via RoundTrip
+	creds, connClosed, err := a.getCreds()
 	if err != nil {
 		return nil, err
 	}
+	if connClosed {
+		// If we close the underlying TCP connection in the process of starting the TLS handshake,
+		// there is no point in letting the handshake continue.  At least this way we provide a
+		// better error message then "write tcp host:port->host:port use of closed network connection"
+		return nil, errors.New("failing TLS handshake due to certificate rotation")
+	}
```

Comment:
I am unsure of a good way to test this path, or if there is some way to avoid this altogether.

Reply:
I could also just turn this into a `klog` statement instead of a direct error.  That way if this logic is triggered in a case that somehow does not result in a `use of closed network connection` error, the TLS handshake would be allowed to continue.

Reply:
I managed to expand `TestTLSCredentialsCallsAndRotation` with some keep-alive shenanigans to test this case.